packages/angular/ssr/src/render.ts

```diff
@@ -0,0 +1,130 @@
+ *
+ * @param app - The server application instance to render.
+ * @param request - The incoming HTTP request object.
+ * @param serverContext - Context specifying the rendering mode.
+ * @param requestContext - Optional additional context for the request, such as metadata.
+ * @returns A promise that resolves to a response object representing the rendered content.
+ */
+export async function render(
+  app: AngularServerApp,
+  request: Request,
+  serverContext: ServerRenderContext,
+  requestContext?: unknown,
+): Promise<Response> {
+  const isSsrMode = serverContext === ServerRenderContext.SSR;
+  const responseInit: ResponseInit = {};
+  const platformProviders: StaticProvider = [
+    {
+      provide: SERVER_CONTEXT,
+      useValue: serverContext,
+    },
+  ];
+
+  if (isSsrMode) {
+    platformProviders.push(
+      {
+        provide: REQUEST,
+        useValue: request,
+      },
+      {
+        provide: REQUEST_CONTEXT,
+        useValue: requestContext,
+      },
+      {
+        provide: RESPONSE_INIT,
+        useValue: responseInit,
+      },
+    );
+  }
+
+  const { manifest, hooks, isDevMode } = app;
+
+  if (isDevMode) {
+    // Need to clean up GENERATED_COMP_IDS map in `@angular/core`.
+    // Otherwise an incorrect component ID generation collision detected warning will be displayed in development.
+    // See: https://github.com/angular/angular-cli/issues/25924
+    ɵresetCompiledComponents();
+
+    // An Angular Console Provider that does not print a set of predefined logs.
+    platformProviders.push({
+      provide: ɵConsole,
+      useCase: Console,
+    });
+  }
+
+  let html = await app.getServerAsset('index.server.html');
+  // Skip extra microtask if there are no pre hooks.
+  if (hooks.has('html:transform:pre')) {
+    html = await hooks.run('html:transform:pre', { html });
+  }
+
+  let url = request.url;
+
+  // A request to `http://www.example.com/page/index.html` will render the Angular route corresponding to `http://www.example.com/page`.
+  if (url.includes('/index.html')) {
+    const urlToModify = new URL(url);
+    urlToModify.pathname = urlToModify.pathname.replace(/index\.html$/, '');
+    url = urlToModify.toString();
+  }
+
+  const bootstrap = manifest.bootstrap();
+  html = isNgModule(bootstrap)
+    ? await renderModule(bootstrap, { url, document: html, extraProviders: platformProviders })
+    : await renderApplication(bootstrap, {
+        url,
+        document: html,
+        platformProviders,
+      });
+
+  return new Response(html, responseInit);
+}
+
+/**
+ * Type guard to determine if a given value is an Angular module.
+ * Angular modules are identified by the presence of the `ɵmod` static property.
+ * This function helps distinguish between Angular modules and bootstrap functions.
+ *
+ * @param value - The value to be checked.
+ * @returns True if the value is an Angular module (i.e., it has the `ɵmod` property), false otherwise.
+ */
+function isNgModule(
```

Comment:
We can potentially use the `reflectComponentType` public API here, which would return `null` for NgModules, see https://angular.dev/api/core/reflectComponentType?tab=usage-notes.

Reply:
I do not think that `reflectComponentType` will work in this case because we do not have a component. What we need is a way to distinguish between a function that returns a promise and an NgModule.